Linker errors

Hello! I've been having fun trying out this project with something I've been working on [here](https://github.com/JaminKoke/meshbool/tree/zngur)!

I've recently just got everything to compile, except now I'm coming across linker errors!

```
cd /home/user0/.local/src/meshbool/cpp/build/test && /usr/bin/cmake -E cmake_link_script CMakeFiles/manifold_test.dir/link.txt --verbose=1
/usr/bin/c++ -Wl,--dependency-file=CMakeFiles/manifold_test.dir/link.d CMakeFiles/manifold_test.dir/test_main.cpp.o CMakeFiles/manifold_test.dir/polygon_test.cpp.o CMakeFiles/manifold_test.dir/properties_test.cpp.o CMakeFiles/manifold_test.dir/manifold_test.cpp.o CMakeFiles/manifold_test.dir/boolean_test.cpp.o CMakeFiles/manifold_test.dir/sdf_test.cpp.o CMakeFiles/manifold_test.dir/smooth_test.cpp.o CMakeFiles/manifold_test.dir/hull_test.cpp.o CMakeFiles/manifold_test.dir/samples_test.cpp.o CMakeFiles/manifold_test.dir/boolean_complex_test.cpp.o CMakeFiles/manifold_test.dir/cross_section_test.cpp.o CMakeFiles/manifold_test.dir/manifoldc_test.cpp.o CMakeFiles/manifold_test.dir/__/samples/src/bracelet.cpp.o CMakeFiles/manifold_test.dir/__/samples/src/knot.cpp.o -o manifold_test   -L/home/user0/.local/src/meshbool/cpp/build  -Wl,-rpath,/home/user0/.local/src/meshbool/cpp/build/samples:/home/user0/.local/src/meshbool/cpp/build/bindings/c:/home/user0/.local/src/meshbool/cpp/build /usr/lib/libgtest_main.so.1.17.0 ../samples/libsamples.so.3.2.1 ../bindings/c/libmanifoldc.so.3.2.1 /usr/lib/libgtest.so.1.17.0 ../libmanifold.so.3.2.1 -lmeshbool ../_deps/clipper2-build/libClipper2.a -lm
/usr/bin/ld: CMakeFiles/manifold_test.dir/test_main.cpp.o: in function `rust::core::option::Option<rust::Ref<float> >::unwrap(rust::core::option::Option<rust::Ref<float> >)':
test_main.cpp:(.text._ZN4rust4core6option6OptionIJNS_3RefIfEEEE6unwrapES5_[_ZN4rust4core6option6OptionIJNS_3RefIfEEEE6unwrapES5_]+0x3c): undefined reference to `unsigned char* rust::__zngur_internal_data_ptr<rust::Ref<float> >(rust::Ref<float> const&)'
/usr/bin/ld: test_main.cpp:(.text._ZN4rust4core6option6OptionIJNS_3RefIfEEEE6unwrapES5_[_ZN4rust4core6option6OptionIJNS_3RefIfEEEE6unwrapES5_]+0x62): undefined reference to `void rust::__zngur_internal_assume_init<rust::Ref<float> >(rust::Ref<float>&)'
/usr/bin/ld: CMakeFiles/manifold_test.dir/smooth_test.cpp.o: in function `rust::core::option::Option<rust::Ref<double> >::unwrap(rust::core::option::Option<rust::Ref<double> >)':
smooth_test.cpp:(.text._ZN4rust4core6option6OptionIJNS_3RefIdEEEE6unwrapES5_[_ZN4rust4core6option6OptionIJNS_3RefIdEEEE6unwrapES5_]+0x3c): undefined reference to `unsigned char* rust::__zngur_internal_data_ptr<rust::Ref<double> >(rust::Ref<double> const&)'
/usr/bin/ld: smooth_test.cpp:(.text._ZN4rust4core6option6OptionIJNS_3RefIdEEEE6unwrapES5_[_ZN4rust4core6option6OptionIJNS_3RefIdEEEE6unwrapES5_]+0x62): undefined reference to `void rust::__zngur_internal_assume_init<rust::Ref<double> >(rust::Ref<double>&)'
collect2: error: ld returned 1 exit status
make[2]: *** [test/CMakeFiles/manifold_test.dir/build.make:319: test/manifold_test] Error 1
make[2]: Leaving directory '/home/user0/.local/src/meshbool/cpp/build'
make[1]: *** [CMakeFiles/Makefile2:1521: test/CMakeFiles/manifold_test.dir/all] Error 2
make[1]: Leaving directory '/home/user0/.local/src/meshbool/cpp/build'
make: *** [Makefile:149: all] Error 2
```

I'm not sure if this is a problem with the way I did things, or if it is a problem inside of zngur.

To reproduce this clone [the repo](https://github.com/JaminKoke/meshbool/tree/zngur), `cd cpp`, `mkdir build && cd build`, `cmake ..`, and `make`.
The header file generated by zngur is created in `generated/meshbool/meshbool.h`. For the Rust side, I have a string substitution in the build script that replaces instances of `#[no_mangle]` with `#[unsafe(no_mangle)]`.

Thank you for your time!

--- a/examples/regression_test1/expected_output.txt
+++ b/examples/regression_test1/expected_output.txt
@@ -1,45 +1,45 @@
 Test dbg works for Ref and RefMut -- started
-[main.cpp:10] v1 = "foo"
-[main.cpp:12] v2 = "foo"
-[main.cpp:14] v3 = "foo"
-[main.cpp:15] v2 = "foo"
-[main.cpp:16] v4 = "foo"
-[main.cpp:18] v5 = "foo"
-[main.cpp:19] "bar"_rs = "bar"
-[main.cpp:20] v4 = "foobar"
+[main.cpp:11] v1 = "foo"
+[main.cpp:13] v2 = "foo"
+[main.cpp:15] v3 = "foo"
+[main.cpp:16] v2 = "foo"
+[main.cpp:17] v4 = "foo"
+[main.cpp:19] v5 = "foo"
+[main.cpp:20] "bar"_rs = "bar"
+[main.cpp:21] v4 = "foobar"
 Test dbg works for Ref and RefMut -- finished
 
 Test fields and constructor work -- started
-[main.cpp:33] v1 = Foo {
+[main.cpp:31] v1 = Foo {
     field1: 1,
     field2: "bar",
 }
-[main.cpp:34] v1.field2 = "bar"
-[main.cpp:35] v1.field2.len() = 3
-[main.cpp:37] v1 = Foo {
+[main.cpp:32] v1.field2 = "bar"
+[main.cpp:33] v1.field2.len() = 3
+[main.cpp:35] v1 = Foo {
     field1: 1,
     field2: "barbaz",
 }
-[main.cpp:40] v2 = (
+[main.cpp:39] v2 = (
     "kkk",
     Foo {
         field1: 1,
         field2: "barbaz",
     },
 )
-[main.cpp:41] v2.f0 = "kkk"
-[main.cpp:42] v2.f1 = Foo {
+[main.cpp:40] v2.f0 = "kkk"
+[main.cpp:41] v2.f1 = Foo {
     field1: 1,
     field2: "barbaz",
 }
-[main.cpp:43] v2.f1.field2 = "barbaz"
-[main.cpp:47] v3.f0 = "kkk"
-[main.cpp:48] v3.f1 = Foo {
+[main.cpp:42] v2.f1.field2 = "barbaz"
+[main.cpp:46] v3.f0 = "kkk"
+[main.cpp:47] v3.f1 = Foo {
     field1: 1,
     field2: "barbazxxx",
 }
-[main.cpp:49] v3.f1.field2 = "barbazxxx"
-[main.cpp:52] v3.f1.field2.len() = 9
+[main.cpp:48] v3.f1.field2 = "barbazxxx"
+[main.cpp:51] v3.f1.field2.len() = 9
 [main.cpp:55] v4.f0 = "kkk"
 [main.cpp:56] v4.f1 = Foo {
     field1: 1,
@@ -50,12 +50,30 @@ Test fields and constructor work -- started
 Test fields and constructor work -- finished
 
 Test Field* underlying conversions -- started
-[main.cpp:70] v0 = 42
-[main.cpp:74] v1 = "hi"
-[main.cpp:78] sref.len() = 2
-[main.cpp:81] int32_t(pref.f0) = 42
-[main.cpp:82] pref.f1.len() = 2
-[main.cpp:85] int32_t(pmut.f0) = 42
-[main.cpp:87] pmut.f1.len() = 3
+[main.cpp:71] v0 = 42
+[main.cpp:75] v1 = "hi"
+[main.cpp:79] sref.len() = 2
+[main.cpp:82] int32_t(pref.f0) = 42
+[main.cpp:83] pref.f1.len() = 2
+[main.cpp:86] int32_t(pmut.f0) = 42
+[main.cpp:88] pmut.f1.len() = 3
 Test Field* underlying conversions -- finished
+
+Test floats -- started
+[main.cpp:98] *r1 = 12.3
+[main.cpp:100] v1 = 12.3
+[main.cpp:105] fvec = [
+    42.24,
+    147.0,
+]
+[main.cpp:106] fvec.get(0) = Some(
+    42.24,
+)
+[main.cpp:107] fvec.get(2) = None
+[main.cpp:108] *fvec.get(1).unwrap() = 147
+[main.cpp:110] fvec = [
+    42.24,
+    5.43,
+]
+Test floats -- finished
 

--- a/examples/regression_test1/main.cpp
+++ b/examples/regression_test1/main.cpp
@@ -4,8 +4,9 @@
 #include "./generated.h"
 
 void test_dbg_works_for_ref_and_refmut() {
-  auto scope = rust::crate::Scoped::new_("Test dbg works for Ref and RefMut"_rs);
-  
+  auto scope =
+      rust::crate::Scoped::new_("Test dbg works for Ref and RefMut"_rs);
+
   rust::Ref<rust::Str> v1 = "foo"_rs;
   zngur_dbg(v1);
   rust::std::string::String v2 = v1.to_owned();
@@ -20,11 +21,8 @@ void test_dbg_works_for_ref_and_refmut() {
   zngur_dbg(v4);
 }
 
-template<typename T>
-concept has_push_str = requires(T v, rust::Ref<rust::Str> s) { 
-  v.push_str(s); 
-};
-
+template <typename T>
+concept has_push_str = requires(T v, rust::Ref<rust::Str> s) { v.push_str(s); };
 
 void test_fields_and_constructor() {
   auto scope = rust::crate::Scoped::new_("Test fields and constructor work"_rs);
@@ -36,13 +34,14 @@ void test_fields_and_constructor() {
   v1.field2.push_str("baz"_rs);
   zngur_dbg(v1);
 
-  rust::Tuple<rust::std::string::String, rust::crate::Foo> v2{"kkk"_rs.to_owned(), std::move(v1)};
+  rust::Tuple<rust::std::string::String, rust::crate::Foo> v2{
+      "kkk"_rs.to_owned(), std::move(v1)};
   zngur_dbg(v2);
   zngur_dbg(v2.f0);
   zngur_dbg(v2.f1);
   zngur_dbg(v2.f1.field2);
   v2.f1.field2.push_str("xxx"_rs);
-  
+
   rust::Ref<rust::Tuple<rust::std::string::String, rust::crate::Foo>> v3 = v2;
   zngur_dbg(v3.f0);
   zngur_dbg(v3.f1);
@@ -51,7 +50,8 @@ void test_fields_and_constructor() {
   static_assert(!has_push_str<decltype(v3.f1.field2)>);
   zngur_dbg(v3.f1.field2.len());
 
-  rust::RefMut<rust::Tuple<rust::std::string::String, rust::crate::Foo>> v4 = v2;
+  rust::RefMut<rust::Tuple<rust::std::string::String, rust::crate::Foo>> v4 =
+      v2;
   zngur_dbg(v4.f0);
   zngur_dbg(v4.f1);
   zngur_dbg(v4.f1.field2);
@@ -60,7 +60,8 @@ void test_fields_and_constructor() {
 }
 
 void test_field_underlying_conversions() {
-  auto scope = rust::crate::Scoped::new_("Test Field* underlying conversions"_rs);
+  auto scope =
+      rust::crate::Scoped::new_("Test Field* underlying conversions"_rs);
 
   rust::Tuple<int32_t, rust::std::string::String> pair{42, "hi"_rs.to_owned()};
 
@@ -87,8 +88,31 @@ void test_field_underlying_conversions() {
   zngur_dbg(pmut.f1.len());
 }
 
+void test_floats() {
+  auto scope = rust::crate::Scoped::new_("Test floats"_rs);
+
+  rust::Tuple<float, double> pair{42.24, 12.3};
+
+  // FieldOwned conversion to Ref and value
+  rust::Ref<double> r1 = pair.f1;
+  zngur_dbg(*r1);
+  double v1 = pair.f1;
+  zngur_dbg(v1);
+
+  rust::std::vec::Vec<float> fvec = rust::std::vec::Vec<float>::new_();
+  fvec.push(pair.f0);
+  fvec.push(147);
+  zngur_dbg(fvec);
+  zngur_dbg(fvec.get(0));
+  zngur_dbg(fvec.get(2));
+  zngur_dbg(*fvec.get(1).unwrap());
+  *fvec.get_mut(1).unwrap() = 5.43;
+  zngur_dbg(fvec);
+}
+
 int main() {
   test_dbg_works_for_ref_and_refmut();
   test_fields_and_constructor();
   test_field_underlying_conversions();
+  test_floats();
 }

--- a/examples/regression_test1/main.zng
+++ b/examples/regression_test1/main.zng
@@ -1,5 +1,10 @@
 #convert_panic_to_exception
 
+type bool {
+	#layout(size = 1, align = 1);
+	wellknown_traits(Copy);
+}
+
 type str {
     wellknown_traits(?Sized, Debug);
 
@@ -39,6 +44,44 @@ type (i32, ::std::string::String) {
 
     field 0 (offset = 0, type = i32);
     field 1 (offset = 8, type = ::std::string::String);
+}
+
+type (f32, f64) {
+    #layout(size = 16, align = 8);
+    wellknown_traits(Debug);
+
+    field 0 (offset = 0, type = f32);
+    field 1 (offset = 8, type = f64);
+}
+
+mod ::std::option {
+    type Option<&f32> {
+        #layout(size = 8, align = 8);
+        wellknown_traits(Debug, Copy);
+
+        fn is_some(&self) -> bool;
+        fn unwrap(self) -> &f32;
+    }
+
+    type Option<&mut f32> {
+        #layout(size = 8, align = 8);
+        wellknown_traits(Debug);
+
+        fn is_some(&self) -> bool;
+        fn unwrap(self) -> &f32;
+    }
+}
+
+mod ::std::vec {
+    type Vec<f32> {
+        #layout(size = 24, align = 8);
+        wellknown_traits(Debug);
+
+        fn new() -> Vec<f32>;
+        fn get(&self, usize) -> ::std::option::Option<&f32> deref [f32];
+        fn get_mut(&mut self, usize) -> ::std::option::Option<&mut f32> deref [f32];
+        fn push(&mut self, f32);
+    }
 }
 
 type crate::Scoped {

--- a/zngur-generator/src/template.rs
+++ b/zngur-generator/src/template.rs
@@ -58,22 +58,21 @@ pub(crate) struct CppHeaderTemplate<'a> {
 impl<'a> CppHeaderTemplate<'a> {
     // TODO: Docs - what do these represent? When will we change this list?
     fn builtin_types(&self) -> Vec<String> {
-        [8, 16, 32, 64]
+        let builtins = [8, 16, 32, 64]
             .into_iter()
             .flat_map(|x| [format!("int{x}_t"), format!("uint{x}_t")])
-            .chain([8, 16, 32, 64].into_iter().flat_map(|x| {
+            .chain(["::double_t".to_owned(), "::float_t".to_owned()])
+            .flat_map(|x| {
                 [
-                    format!("::rust::Ref<int{x}_t>"),
-                    format!("::rust::Ref<uint{x}_t>"),
-                    format!("::rust::RefMut<int{x}_t>"),
-                    format!("::rust::RefMut<uint{x}_t>"),
+                    x.clone(),
+                    format!("::rust::Ref<{x}>"),
+                    format!("::rust::RefMut<{x}>"),
                 ]
-            }))
+            });
+        builtins
             .chain([
-                "::rust::ZngurCppOpaqueOwnedObject".to_string(),
-                "::double_t".to_string(),
-                "::float_t".to_string(),
-                "::size_t".to_string(),
+                "::rust::ZngurCppOpaqueOwnedObject".to_owned(),
+                "::size_t".to_owned(),
             ])
             .collect()
     }

